.github/workflows/publish.yml

```diff
@@ -0,0 +1,47 @@
+name: Publish to npm
+on:
+  push:
+    branches:
+      - master
+    paths:
+      - 'package.json'
+jobs:
+  test:
+    uses: ./.github/workflows/test.yml
+  get-version:
+    runs-on: ubuntu-latest
+    outputs:
+      current_version: ${{ steps.current_version.outputs.version }}
+      last_version: ${{ steps.last_version.outputs.version }}
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 50
+      # Check if the package.json version field has changed since the last push
+      - name: Get current version from package.json
+        id: current_version
+        run: |
+          echo "version=$(jq -r .version package.json)" >> $GITHUB_OUTPUT
+      - name: Get the version from the last push
+        id: last_version
+        run: |
+          git checkout ${{ github.event.before }}
+          echo "version=$(jq -r .version package.json)" >> $GITHUB_OUTPUT
+  npm-publish:
+    needs:
+      - test
+      - get-version
+    runs-on: ubuntu-latest
+    # We only want to publish if the package.json version field has changed
+    if: needs.get-version.outputs.current_version != needs.get-version.outputs.last_version
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-node@v3
+        with:
+          node-version: 14
```

Comment:
Does the package lock in the repository not work under Node 14?
Anyway, our `tsconfig.json` should tell the build system to target es2020 (which is compatible with Node 14) regardless of the current Node version: https://github.com/smogon/pokemon-showdown/blob/70759c5c13ff1d4ea56ee63d44a5202e67f20eee/tsconfig.json#L2-L3

Reply:
> Does the package lock in the repository not work under Node 14?

I don't think so. I tried on my local machine via nvm and it fails, but 19 is fine. If it's ok to change that I'll update the CI to use a modern version of node

Reply:
npm v6 or lower cant parse the lockVersion v2 or higher, should install by `npm i`

Reply:
I've updated to a modern version of node which fixes this

Reply:
anyway v19 will be deprecated soon, use v16 instead

Reply:
Are you sure? v16 is under maintenance. I do agree v19 might not be the best idea after checking it, but isn't v18 the best one to use for LTS?

Reply:
![image](https://user-images.githubusercontent.com/6253896/227812824-974495a3-46a7-4fbd-a584-71819dc61f8f.png)
![image](https://user-images.githubusercontent.com/6253896/227812834-7be95658-dd5f-4f11-8d4a-6097933f9db1.png)

Reply:
uh i thought it ends in 2024
but Zarel didn't say using v18, please wait his response, in any case it is not recommended using odd number version.

Reply:
Gotcha, yeah I agree we probably shouldn't use v19. I recently upgraded some other workflows of mine to node 19 based on a dependabot recomendation, however I didn't actually check when supported ended. I assumed it would last a bit longer than this.

In theory we can use any version for the CI, as long as the build works. Developers are free to use any version on their local machines

Just going to @Zarel to confirm what version we should use for the CI

Reply:
Okay, yes, I'd prefre 18.